tests/functions/test_expand_grid.py

```diff
@@ -0,0 +1,635 @@
+    }
+
+    #collect dataframes here
+    dfs = []
+
+    #collect non dataframes here, proper dicts ... key value pair where the value is a list of scalars
+    dicts = {}
+
+    for key, value in entry.items():
+
+        #exclude dicts:
+        if isinstance(value, dict):
+            raise ValueError("nested dicts not allowed")
+
+        #process arrays
+        if isinstance(value, np.ndarray):
+            if value.size == 0:
+                raise ValueError("array cannot be empty")
+            elif value.ndim == 1:
+                dfs.append(pd.DataFrame(value, columns=[key]))
+            elif value.ndim == 2:
+                dfs.append(pd.DataFrame(value).add_prefix(f"{key}_"))
+            else:
+                raise ValueError(
+                    "expand_grid works with only vector and matrix arrays")
+        #process series
+        if isinstance(value, pd.Series):
+            if value.empty:
+                raise ValueError("passed Series cannot be empty")
+            if not isinstance(value.index, pd.MultiIndex):
+                if value.name:
+                    value = value.to_frame(name=f"{key}_{value.name}")
+                    dfs.append(value)
+                else:
+                    value = value.to_frame(name=f"{key}")
+                    dfs.append(value)
+            else:
+                raise ValueError(
+                    "expand_grid does not work with pd.MultiIndex")
+        #process dataframe
+        if isinstance(value, pd.DataFrame):
+            if value.empty:
+                raise ValueError("passed DataFrame cannot be empty")
+            if not (isinstance(value.index, pd.MultiIndex)
+                    or isinstance(value.columns, pd.MultiIndex)):
+                #add key to dataframe columns
+                value = value.add_prefix(f"{key}_")
+                dfs.append(value)
+            else:
+                raise ValueError(
+                    "expand grid does not work with pd.MultiIndex")
+        #process lists
+        if isinstance(value, list):
+            if not value:
+                raise ValueError("passed value cannot be empty")
+            elif np.array(value).ndim == 1:
+                checklist = (type(None), str, int, float, bool)
+                check = (isinstance(internal, checklist) for internal in value)
+                if all(check):
+                    dicts.update({key: value})
+                else:
+                    raise ValueError("values in iterable must be scalar")
+            elif np.array(value).ndim == 2:
+                value = pd.DataFrame(value).add_prefix(f"{key}_")
+                dfs.append(value)
+            else:
+                raise ValueError("sequence's dimension should be 1d or 2d")
+
+    return dfs, dicts
+
+
+#computation for values that are not arrays/dataframes/series
+#these are collected into dictionary and processed with numpy meshgrid
+def grid_computation_dict(dicts):
```

Comment:
I believe the same comment as above applies here.

Copy/pasting it for immediate visible context: This function should be imported from the library, rather than re-defined here.